fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/IndexAggregateFunctionCall.java

```diff
@@ -0,0 +1,272 @@
+/*
+ * IndexAggregateFunctionCall.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.metadata;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.metadata.expressions.EmptyKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.GroupingKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.query.expressions.AndComponent;
+import com.apple.foundationdb.record.query.expressions.BaseField;
+import com.apple.foundationdb.record.query.expressions.Comparisons;
+import com.apple.foundationdb.record.query.expressions.FieldWithComparison;
+import com.apple.foundationdb.record.query.expressions.NestedField;
+import com.apple.foundationdb.record.query.expressions.QueryComponent;
+import com.apple.foundationdb.record.query.plan.temp.EnumeratingIterable;
+import com.apple.foundationdb.record.query.plan.temp.TopologicalSort;
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * Class to represent a construct {@code AGG by Set.of(col1, col2, ...)}. This is the unbound equivalent of
+ * {@link IndexAggregateFunction} which is bound to an index and represents its group-by columns as a list of
+ * {@link KeyExpression} that match the order imposed by bound index rather than a set.
+ */
+@SuppressWarnings({"unused", "OptionalUsedAsFieldOrParameterType"})
+@API(API.Status.INTERNAL)
+public class IndexAggregateFunctionCall {
+    @Nonnull
+    private final String name;
+    @Nonnull
+    private final GroupingKeyExpression groupingKeyExpression;
+    @Nonnull
+    private final Set<KeyExpression> groupingExpressions;
+    @Nonnull
+    private final Optional<KeyExpression> groupedExpressionOptional;
```

Comment:
No (I think), because `EmptyKeyExpression` is still something in terms of other methods using the result of this down the road. The reason I added the `Optional<>` was because I didn't want to presume that there was a no legitimate reason for empty to occur _naturally_. Then again, I don't know whether the original reason for empty in the code base was to actually represent the absence of a key expression or a key expression that returns an empty.
Also, optionals feel more natural to me to indicate the absence of something and nicely compose functionally. Let me know what you want me to do here.

Reply:
> the absence of a key expression or a key expression that returns an empty

What is the practical difference? Returning empty means contributing zero columns to the evaluation.

Is the code that would make a distinction using this new class yet to be written? I don't see anything below that wouldn't just work.

Reply:
For instance, if I want to create a GroupingKeyExpression from it and some grouping keys, I can easily say:

`new GroupingKeyExpression(concat(groupingKey, groupedKey), groupedKey.getColumnSize())`

That's fine if it is not `EmptyKeyExpression`; if it is you probably want to say something like

`new GroupingKeyExpression(groupingKey, 0)`

Now regardless of whether `groupedExpression` is an `EmptyKeyExpression` or an `Optional<KeyExpression>` the code needs to make this distinction (case 1 or 2) with the difference that the caller is forced to think about it from a typing perspective: what happens if the key expression is not there? If it is just the ``EmptyKeyExpression` I can write something like: `new GroupingKeyExpression(concat(groupingKey, groupedKey), groupedKey.getColumnSize())`
even though I may not intend to. And this is exactly what happened downstream where I messed up an index match because further down I got an `EmptyKeyExpression` within a bunch of other keys after some other manipulations.

Reply:
Resolves off-thread: Will change implementation to use `EmptyKeyExpression` mostly due to convention.